cfme/tests/integration/test_cfme_auth.py

```diff
@@ -1,90 +1,156 @@
+           # need SSL openldap server
+           unblock=lambda auth_mode, prov_key: not (
+               auth_mode in ['external', 'ldaps'] and
+               auth_data.auth_providers[prov_key].type == 'openldaps')
+           )
+    ])
+]
 
-RETRIEVE_GROUP = 'retrieve_group'
-CREATE_GROUP = 'create_group'
+# map auth provider types, auth_modes, and user_types for test matrix
+# first key level is auth mode
+# second key level is provider type  (auth_provider key in parametrization)
+# finally, user_types valid for testing on the above combination of provider+mode
+test_param_maps = {
+    'amazon': {
+        AmazonAuthProvider.auth_type: {
+            'user_types': ['username']}
+    },
+    'ldap': {
+        ActiveDirectoryAuthProvider.auth_type: {
+            # add cn_domain, samacct
+            'user_types': ['cn', 'email', 'uid', 'upn']
+        },
+        FreeIPAAuthProvider.auth_type: {
+            'user_types': ['cn', 'uid']  # add cn_domain
+        },
+        OpenLDAPAuthProvider.auth_type: {
+            'user_types': ['cn', 'uid']  # add cn_domain
+        }
+    },
+    'external': {
+        FreeIPAAuthProvider.auth_type: {
+            'user_types': ['uid']
+        },
+        OpenLDAPSAuthProvider.auth_type: {
+            'user_types': ['uid']
+        }
+        # TODO add ActiveDirectory SAMAcct usertype for external
+    }}
 
 
 def pytest_generate_tests(metafunc):
-    auth_modes = auth_data.get('auth_providers')
-    if not auth_modes:
-        return
-    argvalues = [[authmode] for authmode in auth_modes]
-    if 'configure_auth' in metafunc.fixturenames:
-        metafunc.parametrize(['auth_mode'], argvalues)
-
-
-@pytest.fixture()
-def data(request, auth_mode, add_group):
-    yaml_data = auth_data['auth_test'].get(auth_mode, {})
-    if add_group == 'evm_default_group':
-        yaml_data['get_groups'] = False
-    return yaml_data
-
-
-@pytest.fixture()
-def group(request, data, auth_mode, add_group, appliance):
-    if not data:
-        pytest.skip("No data spcified for user group")
-    credentials = Credential(
-        principal=data["username"],
-        secret=data["password"],
-    )
+    """ zipper auth_modes and auth_prov together and drop the nonsensical combos """
+    argnames = ['auth_mode', 'prov_key', 'user_type']
+    argvalues = []
+    for mode in test_param_maps.keys():
+        for auth_type in test_param_maps.get(mode, {}):
+            eligible_providers = {key: prov_dict
+                                  for key, prov_dict in iteritems(auth_data.auth_providers)
+                                  if prov_dict.type == auth_type}
+            for user_type in test_param_maps[mode][auth_type]['user_types']:
+                argvalues.extend([[mode, key, user_type]
+                                 for key, prov_dict in iteritems(eligible_providers)
+                                 if user_type in prov_dict.get('user_types', [])])
+    metafunc.parametrize(argnames, argvalues)
```

Comment:
**REQUIRED:** No explicit setting of idlist?

Reply:
The argvalues are usable as id's, but yeah I'll explicitly pass here.